### PR TITLE
Enable mesa GLES support for Generic/RPi/RPi2

### PR DESCRIPTION
**Pull request by lrusak**
This is in regards to my chat with @anholt. Unfortunately Kodi doesn't support a generic gles implementation (yet) so it is not possible to run kodi without xorg when using the vc4 driver. This is purely for testing.

When building RPi/RPi2 you will need to use the options
```
OPENGLES="mesa"
GRAPHIC_DRIVERS="vc4"
```
and during runtime you need the following in config.txt
```
dtoverlay=vc4-kms-v3d
```

Not something for merging (for a while)

```
[    4.545288] vc4-drm soc:gpu: bound 3f902000.hdmi (ops 0x8073d5b8)
[    4.545335] vc4-drm soc:gpu: bound 3f400000.hvs (ops 0x8073d860)
[    4.545523] vc4-drm soc:gpu: bound 3f206000.pixelvalve (ops 0x8073c9ec)
[    4.545663] vc4-drm soc:gpu: bound 3f207000.pixelvalve (ops 0x8073c9ec)
[    4.545799] vc4-drm soc:gpu: bound 3f807000.pixelvalve (ops 0x8073c9ec)
[    4.545850] vc4-drm soc:gpu: bound 3fc00000.v3d (ops 0x8073dc24)
[    4.553041] [drm] Supports vblank timestamp caching Rev 2 (21.10.2013).
[    4.553057] [drm] Driver supports precise vblank timestamp query.
[    4.741072] vc4-drm soc:gpu: fb0:  frame buffer device
```

using the libdrm modetest utility everything seems to be working properly
```
LibreELEC:~ # ./modetest -M vc4
Encoders:
id	crtc	type	possible crtcs	possible clones	
23	48	TMDS	0x00000004	0x00000000

Connectors:
id	encoder	status		name		size (mm)	modes	encoders
24	23	connected	HDMI-A-1       	520x300		22	23
<snip>
```

### CODE REVIEW

**Comment by MilhouseVH** [comment]
Can this be rebased on master (with 4.10.0 kernel)?

**Comment by anholt** [comment]
A few thoughts:

- The kernel config changes should be safe to include regardless of whether you're using the closed or open drivers.  The overlay will decide if the DRM module will actually load.
- Are you using libkms for something?  I'm surprised you're enabling it in the vc4 case in libdrm, since it shouldn't be any more needed on vc4 than it is on i965.
- 7b8b021df23ec434b1e7f4121cbcfef7fde11892 was hard to follow because there were a bunch of changes that didn't seen to be related to vc4.

**Comment by lrusak** [comment]
@MilhouseVH updated

@anholt are you saying we don't need to enable libkms for any of the drivers?

Yea the mesa commit is pulled from my wayland branch where I had to support building mesa for x11 and wayland so in our case opengl vs opengles. Currently only x86 uses mesa at all but with these changes we can build mesa for arm (RPi) as well. 

So to sum it up when building mesa without x11 the opengles libraries will be built as well as the selected gallium driver.

**Comment by anholt** [comment]
I don't know of any reason for needing libkms, unless software you've built on top is using it.

**Comment by InuSasha** [comment]
I tried to build this PR for RPi2, but it was failing.
The X11/Xlib.h was missing, in the build of kodi:target.
```
In file included from /home/inu/LibreElec/Master/build.LibreELEC-RPi2.arm-9.0-devel-vc4/toolchain/armv7ve-libreelec-linux-gnueabi/sysroot/usr/include/EGL/egl.h:39:0,
                 from /home/inu/LibreElec/Master/build.LibreELEC-RPi2.arm-9.0-devel-vc4/kodi-4c792db/xbmc/windowing/egl/WinSystemEGL.h:31,
                 from /home/inu/LibreElec/Master/build.LibreELEC-RPi2.arm-9.0-devel-vc4/kodi-4c792db/xbmc/windowing/WindowingFactory.h:45,
                 from /home/inu/LibreElec/Master/build.LibreELEC-RPi2.arm-9.0-devel-vc4/kodi-4c792db/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp:35:
/home/inu/LibreElec/Master/build.LibreELEC-RPi2.arm-9.0-devel-vc4/toolchain/armv7ve-libreelec-linux-gnueabi/sysroot/usr/include/EGL/eglplatform.h:119:22: fatal error: X11/Xlib.h: No such file or directory
 #include <X11/Xlib.h>
```
Full-Log: https://dl.inusasha.de/libreelec/logs/build_pr1395_vc4_2017.03.04.log.gz

Changed RPi2/options:
```diff
diff --git a/projects/RPi2/options b/projects/RPi2/options
index b54cc79a0..c6caf5f74 100644
--- a/projects/RPi2/options
+++ b/projects/RPi2/options
@@ -87,7 +87,7 @@
     OPENGL="no"
 
   # OpenGL-ES implementation to use (no / bcm2835-driver / gpu-viv-bin-mx6q)
-    OPENGLES="bcm2835-driver"
+    OPENGLES="mesa"
 
   # include uvesafb support (yes / no)
     UVESAFB_SUPPORT="no"
@@ -101,7 +101,7 @@
   # Xorg Graphic drivers to use (all / i915,i965,r200,r300,r600,nvidia)
   # Space separated list is supported,
   # e.g. GRAPHIC_DRIVERS="i915 i965 r300 r600 radeonsi nvidia"
-    GRAPHIC_DRIVERS=""
+    GRAPHIC_DRIVERS="vc4"
 
   # KODI Player implementation to use (default / bcm2835-driver / libfslvpuwrap)
     KODIPLAYER_DRIVER="bcm2835-driver"
```

**Comment by lrusak** [comment]
@InuSasha that's to be expected. It will require some extra cflags. Something like this will need to be added to Kodi, however, Kodi cannot run with a Generic OpenGLES implementation right now so it's useless.

```
  elif [ "$KODIPLAYER_DRIVER" = mesa ]; then
    KODI_PLAYER="-DCORE_SYSTEM_NAME=linux"
    CFLAGS="$CFLAGS -DMESA_EGL_NO_X11_HEADERS"
    CXXFLAGS="$CXXFLAGS -DMESA_EGL_NO_X11_HEADERS"
```

**Comment by InuSasha** [comment]
@lrusak ok. I thought this PR would be a little further. My mistake.
will take a look, what will change with the new flags :)

**Comment by lrusak** [comment]
@InuSasha you could try just using opengl and x11, though I've never tried building x11 on the RPi.

**Comment by InuSasha** [comment]
@lrusak OpenGL+X11 builds, but does actual not run (xorg-server announce maybe unsupported driver). Addional, i think it is the wrong way, because opengl is not supported by the hardware, like opengles.

all tested variants with opengles via mesa, was failing (problem with the native opengles classes in kodi)

**Comment by popcornmix** [comment]
> Addional, i think it is the wrong way, because opengl is not supported by the hardware, like opengles.

OpenGL has hardware support (for almost all operations) with the vc4 driver

**Comment by InuSasha** [comment]
@popcornmix: thanks for correction

**Comment by InuSasha** [comment]
@lrusak i got kodi running with x11+opengl via mesa.
But i had to start "/usr/lib/xorg/xorg-configure" manuell, via ssh.

i tries to see some videos. SD performance is good. HD performance is not. every second is a short stop.

**Comment by lrusak** [comment]
I've done work towards getting Kodi running on DRM/KMS without X11. Still a long way to go but it's something

[Imgur](http://i.imgur.com/BSLBWKV.jpg)

**Comment by lrusak** [comment]
@anholt do you know how the performance is on the RPi2 with the vc4 driver? I seem to run out of buffer objects and I'm not sure how to fix that other than just dropping the frame. Maybe you'll have some ideas about this. see, https://github.com/xbmc/xbmc/compare/master...lrusak:drm-kms#diff-c3785d85101925cea9780e262d506886R128

**Comment by anholt** [comment]
vc4 should be faster than closed -- anywhere that it isn't is a bug, basically, and on glmark2 I'm crushing the closed driver.

I'm skeptical of the framebuffer management in your linked KMS code: you'd be better off with X11, since you're immediately blocking on the flip completing, instead of blocking only if you have a new frame ready before the last flip is done.

You can cat /debug/dri/0/bo_stats to see how much BO memory has been allocated in vc4.  From traces I've looked at of Kodi, there's not that much memory being allocated, so hitting the 256MB limit seems more like there's something wrong in the KMS port.

**Comment by lrusak** [comment]
Updated to what I was building with my kodi drm-kms branch. This isn't safe to merge yet.

**Comment by lrusak** [comment]
I've updated this. It should be "safe" to merge. @MilhouseVH can you please try to build with this just your normal builds? I want to make sure this doesn't break anything.

File: packages/graphics/mesa/package.mk [hunk 1]
```diff
@@ -22,14 +22,29 @@ PKG_ARCH="any"
...
 PKG_SECTION="graphics"
 PKG_SHORTDESC="mesa: 3-D graphics library with OpenGL API"
 PKG_LONGDESC="Mesa is a 3-D graphics library with an API which is very similar to that of OpenGL*. To the extent that Mesa utilizes the OpenGL command syntax or state machine, it is being used with authorization from Silicon Graphics, Inc. However, the author makes no claim that Mesa is in any way a compatible replacement for OpenGL or associated with Silicon Graphics, Inc. Those who want a licensed implementation of OpenGL should contact a licensed vendor. While Mesa is not a licensed OpenGL implementation, it is currently being tested with the OpenGL conformance tests. For the current conformance status see the CONFORM file included in the Mesa distribution."
 
 PKG_IS_ADDON="no"
 PKG_AUTORECONF="yes"
 
+if [ "$DISPLAYSERVER" = "x11" ]; then
+  PKG_DEPENDS_TARGET="toolchain Python:host expat glproto dri2proto presentproto libdrm libXext libXdamage libXfixes libXxf86vm libxcb libX11 systemd dri3proto libxshmfence libressl"
```
Comment (MilhouseVH) [inline packages/graphics/mesa/package.mk]: `libressl` dependency? :)

**Comment by MilhouseVH** [comment]
I built this for RPi3, and all is fine without `dtoverlay=vc4-kms-v3d`, however once this overlay is enabled then the LE splash screen framebuffer disappears, and kodi crashes on startup.

Various logs after adding `dtoverlay=vc4-kms-v3d`:

`dmesg`: http://sprunge.us/YZbM
`config.txt`: http://sprunge.us/MOWU
`vcdbg log msg`: http://sprunge.us/JGiF (unfortunately it's like this withing a few seconds of powering on)
`Kodi crashlog`: http://sprunge.us/bYNK

I'm using latest RPi firmware, latest Kodi master (c2ae888).

**Comment by MilhouseVH** [comment]
Here's a kodi crashlog from a debug-enabled build: http://sprunge.us/NTOT

**Comment by lrusak** [comment]
squashed

**Comment by InuSasha** [comment]
packages/mediacenter/kodi/patches/kodi-100.30-rpi-egl.patch does not apply, after merge with master

**Comment by lrusak** [comment]
@InuSasha whoops, those patches aren't even supposed to be include. Thanks

**Comment by lrusak** [comment]
@MilhouseVH any issues?

**Comment by MilhouseVH** [comment]
I haven't been including it in the test builds as crashes and apparently this is expected behaviour so I kind of didn't see the point including this for testing. What is it you need/expect?
